### PR TITLE
Reducers that can listen to other model actions

### LINKED ISSUE #231: Is it possible to have models act on actions within other model
Based on the API I couldn't find if this is possible. Scenario could be as follow:

I have a generic Loading Model that every time I load data (`reducers` or `effects` getting called with naming convention) it should change its state.

I have a User Model and a Project Model. Both of them implement the effect `load` and both of them implement the reducer `loaded` and `loadingFailed`.

### PR DESCRIPTION
**Pull request by ShMcK**
Closes #231. A question turned into a proposal.

Listeners should be able to capture actions for other models directly within reducers.

The implementation can change, it is still up for discussion and debate, just getting the ball rolling.

### CODE REVIEW

**Comment by ShMcK** [comment]
Now a working implementation. Feel free to give it a try. See the test for an example.

**Comment by ShMcK** [comment]
This will also lead to a solid update in the loading plugin.

Anything you might consider changing? Otherwise, will merge tomorrow.

**Comment by fkrauthan** [comment]
Looks very good. The only other thing I was thinking about (but it might be overkill) to allow wildecards. E.g. `'*': () => ` listens to ALL events and reducers and `'test/*': () => ` listens to all events and reducers within the `test` model. That way you can probably turn the loader plugin into a 5 line project.

**Comment by ShMcK** [comment]
Hmm... Can you provide a use case for wild cards?

The logic is already written within the subscriptions plugin, but I haven't found a use for it in my own code. It's also more costly to pattern match on every action dispatched.

**Comment by fkrauthan** [comment]
I think the load plugin is a great use-case for it. Another use case could be a generic error tracer for example listening for `*/*error*` to remove some performance issues you could preprocess the model and only use pattern match when it contains `*`. But to be honest I don't have a full real usecase in mind it just popped into my head when looking over it.

**Comment by ShMcK** [comment]
On that train of thought, it might be a good idea to have a more formalized way of handling errors. 
Worth putting some thought/research into.

**Comment by fkrauthan** [comment]
Oh I just noticed that syntax might also be useful for `effects`. That I can have my own effect execute when another effect in a different model is called.

**Comment by ShMcK** [comment]
TODO:
- [ ] handle dispatches from other model's effects

File: docs/api.md [hunk 1]
```diff
@@ -117,6 +117,16 @@ An object of functions that change the model's state. These functions take the m
 }
 ```
 
+Reducers may also listen to actions from other models by listing the 'model name' + 'action name' as the key.
+
+```js
+{
+  reducers: {
+    'otherModel/actionName': (state, payload) => state + payload,
```
Comment (crazy4groovy) [inline docs/api.md]: Technically, isn't this the "action type"?

Comment (ShMcK) [reply docs/api.md]: Yes. In this case action types are all namespaced by the model name.

And you're right, it's not very clear. How should we phrase this to make it clearer?

Comment (crazy4groovy) [inline docs/api.md]: This is a cool concept!

File: test/listener.test.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      reducers: {
+        'count1/increment': state => state + 1
+      }
+    }
+
+    const store = init({
+      models: { count1, count2 }
+    })
+
+    dispatch.count1.increment()
```
Comment (crazy4groovy) [inline test/listener.test.js]: You may wish to pass in a payload here, and test that it comes through
`'count1/increment': (state, payload) => state + payload`

**Comment by ShMcK** [comment]
I think will publish this as is. 

Multiple "effect" listeners will take a little longer to ensure they're 100%.

Multiple effects can be chained together, no issue there. but must handle several situations:

1. the root effect doesn't exist
2. the root effect is loaded after a listener effect is loaded
3. the listener effect is loaded after a root effect

_(Consider the effect on the model to be the "root effect")_

The order of effects fired is likely important. I'd like to see a solution that ensures the root effect fires first.

**Comment by mgutz** [comment]
Must say this feature was the only thing holding me back. As soon as I saw this framework I thought, "cool but..." This framework has made redux fun instead of being tedious. Great job. Using it on a non-trivial project and so far so good.

**Pull request merged**